Refactoring sample from dicts to named arguments + **kwargs

gvegayon opened the pull request:
This PR includes the following:

- Updates the metaclasses `RandomVariable` and `Model` to use `**kwargs` for `sample()`.
- Updates all instances of these under the submodules.
- Within the `model` module, it ensures that arguments are via keywords instead of dictionaries.
- Updates the tests to reflect the change.
- Updates the demos under docs to reflect the change.

**For a new discussion**: Now that I've done this refactoring, the need to standardize the submodules became apparent. 

- When specifying the observation process for any of the models, it makes sense to have the `RandomVariable.sample()` method of `observed` class feature `mean` and `obs`. This way the `HospitalizationModel` would always expect the observation process to receive `mean` for the latent hospitalizations and `obs` for the observed hospitalizations.
- Other arguments could still be passed as now `**kwargs` are unpacked whenever calling a `RandomVariable.sample()` method.
- Like the `observed`, `process` could also have defaults for `RandomVariable.sample()`, e.g., `n_timesteps`.

gvegayon commented on model/src/pyrenew/deterministic/deterministic.py:
Great point, @AFg6K7h4fhy2! I'll make a note on the metaclass. The idea is to have it as a standard. All methods should have `**kwargs`. This will ensure developers (us) that sample functions of `RandomVariable` and `Model` will always accept `**kwargs` (and thus must be passed through).

dylanhmorris replied:
Tip: you can comment on individual lines of a file by clicking the blue plus button on mouseover

dylanhmorris commented on model/src/pyrenew/deterministic/deterministic.py:
```diff
@@ -41,18 +41,14 @@ def validate(vars: tuple) -> None:
-        constants : dict
-            Ignored. Default None.
+        kwargs : dict
+            Ignored.
```
Agree with @AFg6K7h4fhy2 that I think it would be best to say something like 
```Python
"""
**kwargs : dict, optional
    Additional keyword arguments passed through to internal `sample()` calls, if any
"""
```

gvegayon replied:
In some cases, the kwargs are explicitly ignored, i.e., not passed to anything. But we keep them for consistency. Should I still say these could be passed? That doesn't seem right.

dylanhmorris replied:
Clarification. As I understand it, the pattern we're planning to use is the following:

```Python
class Example(RandomVariable):
   
    def __init__(self):
           self.internal_rv_ = OtherRandomVariable()

    def sample(self, mandatory_named_arg: ArrayLike = None, **kwargs):
           self.do_something_with_named_arg(mandatory_named_arg)
           self.interval_rv_.sample(**kwargs) 
```

So any `RandomVariable.sample()` method only refers to named arguments within its own function body, but it also passes any additional `**kwargs` it receives onward to any calls to other `RandomVariable.sample() methods` that happen under the hood.

dylanhmorris commented on model/src/pyrenew/model/hospitalizations.py:
```diff
@@ -62,7 +64,7 @@ def __init__(
             I0=I0,
             latent_infections=latent_infections,
-            observed_infections=None,
+            observed_infections=DeterministicVariable((0,)),
```
[Mutable default](https://docs.python-guide.org/writing/gotchas/)

gvegayon replied:
But that's within the function's body, not the definition, so it should be OK, right? Also, is it everything a mutable default?  How about strings?

gvegayon replied:
I think the code is OK, @dylanhmorris

gvegayon replied:
Also, I changed a couple of instances where a `RandomVariable` had a default value to instead have None. Those cases with default strings I left them as strings are inmutable.

dylanhmorris replied:
Oops, I was tricked by how GH collapsed the diffs.

dylanhmorris commented on model/src/pyrenew/model/hospitalizations.py:
```diff
@@ -76,23 +78,24 @@ def __init__(
+            **kwargs,
         )
 
     def sample_hospitalizations_obs(
```
Needs updated docstring (still refers to `constants` and `random_variables`)

gvegayon replied:
You are commenting on the deleted code.

dylanhmorris replied:
Oops, sorry!

gvegayon replied:
This was addressed in 44cb3d9

dylanhmorris reviewed (changes requested):
Some tweaks.

RE: `Observation` classes: as discussed in person, we should standardize the name of the predicted value and observed value keyword arguments to an `Observation`'s `sample()` method. I think `obs` for observed (to be consistent with `numpyro.sample()`. For the prediction, how about `pred`, which is nicely consistent with `obs`?

gvegayon commented:
@AFg6K7h4fhy2 and @dylanhmorris, all done with your requests:

- Improved doc on `**kwargs`: I still left a couple where it says `Ignored`, but that's because it is literally that. 
- Renamed `mean` to `predicted` for the observation processes.
- Replaced a default value `RandomVariable` to none to avoid mutable defaults (but left a few strings, as strings are inmutable).

dylanhmorris commented on model/docs/pyrenew_demo.md:
````diff
@@ -93,7 +88,7 @@ hospmodel = HospitalizationsModel(
 ``` python
 with seed(rng_seed=np.random.randint(1, 60)):
-    x = hospmodel.sample(constants=dict(n_timepoints=30))
+    x = hospmodel.sample(n_timepoints=30)
````
Good point. No reason not to say something like `Suppose our timestep size is one day; here's how we'd simulate about a month (30 days) worth of data` in the docs.

gvegayon replied:
This is something that will be relevant for #47 (cc @cshelley). Tagging to be addressed then.

dylanhmorris replied:
Sounds good. @AFg6K7h4fhy2 make an issue and assign yourself / @cshelley ?

dylanhmorris commented on model/src/pyrenew/observation/poisson.py:
```diff
@@ -23,40 +22,36 @@ def __init__(
         self,
-        random_variables: dict,
-        constants: dict = None,
+        predicted: ArrayLike,
+        obs: ArrayLike = None,
```
Update: #65 and #66

gvegayon commented on model/src/pyrenew/process/simplerandomwalk.py:
Keeping that for now, as it is really ignored.

gvegayon commented on model/src/pyrenew/process/rtrandomwalk.py:
Can you point to a specific case? I think most are `<arg> : <type>` (which is what numpydoc does).

gvegayon replied:
Also, I think this kind of thing will be addressed once #38 is up and running!
